### PR TITLE
feat(autodiscover)!: move autodiscovery filter from string to array

### LINKED ISSUE #8763: Make autodiscover filter more useful
**What would you like Renovate to be able to do?**

Rel: #8761 (not a bug)

In original issue I had a problem with brace expansion and forward slash.

**Did you already have any implementation ideas?**

1. Implement filter as or-array
    ```json
    autodiscoverFilter: [
      "@(example1|example2)/*",
      "example3/example4/*",
      "example5"
    ]
    ````
    `--autodiscover-filter="@(example1|example2)/*" --autodiscover-filter="example3/example4/*" --autodiscover-filter="example5"`
1. Add regexp prefix
    `--autodiscover-filter="r/(example1|example2)/i"` - prefix `r` for regexp, regexp flag suffix `i` for case-insensitive
1. Add autodiscovery filter regexp
    `--autodiscover-filter-regexp="/(example1|example2)/i"`
    ```json
    autodiscoverFilterRegexp: "/(example1|example2)/i"
    ```
1. Implement or-array, add regexp support at the same time.

### PR DESCRIPTION
**Pull request by secustor**
## Changes
Change type of `autodiscoverFilter` from string to array. 
 

## Context
- closes #8763
- Supersedes #13100

This is a breaking change which affects administrators running a self hosted instance which uses CLI or ENV for filter configuration and uses `,`. 
e.g. this forces admins which uses minimatches like this `project/{a,b,c,d,*-app}` to migrate to regex ( `/project/(a|b|c|d|.*-app)/` ) or to use a config file. 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/options/index.ts
Comment (HonkingGoose) [inline lib/config/options/index.ts:708]: Do we want a config migration for the change from a `string` to `array`?

Comment (secustor) [reply lib/config/options/index.ts:708]: I think this is ok, as this is a breaking change and we cannot guarantee that we can migrate all config variations because of the massaging.     
But maybe the others have another opinion.

Comment (rarkins) [reply lib/config/options/index.ts:708]: No migrations feasible in this case, we need to document it

File: lib/config/options/index.ts
Comment (viceice) [inline lib/config/options/index.ts:710]: @HonkingGoose This will automigrate to array for config files 😉

File: lib/workers/global/autodiscover.ts
Comment (viceice) [inline lib/workers/global/autodiscover.ts:26]:
```suggestion
  let discovered = await platform.getRepos();
  if (!discovered?.length) {
```

rename not required

File: lib/workers/global/autodiscover.ts
Comment (viceice) [inline lib/workers/global/autodiscover.ts:34]:
```suggestion
```

File: lib/workers/global/autodiscover.ts
Comment (viceice) [inline lib/workers/global/autodiscover.ts:37]:
```suggestion
    discovered = applyFilters(allRepos, config.autodiscoverFilter);
```

File: lib/workers/global/autodiscover.ts
Comment (viceice) [inline lib/workers/global/autodiscover.ts:45]:
```suggestion
```

File: lib/workers/global/autodiscover.ts
Comment (viceice) [inline lib/workers/global/autodiscover.ts:95]: I think we neeed to remove repo from `matched` when a negated regex is in use? eg 

```js
[ 'my-group/**', '!/my-group\\/my-sub-group\\/.*/']
```
This won't work as expected in
- #16466

File: lib/workers/global/config/parse/env.ts
Comment (viceice) [inline lib/workers/global/config/parse/env.ts:168]: this should be a seperate PR

File: lib/workers/global/config/parse/file.ts
Comment (viceice) [inline lib/workers/global/config/parse/file.ts:83]: same here

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts
Comment (HonkingGoose) [reply lib/config/options/index.ts:710]: Cool! 😄

File: lib/workers/global/autodiscover.ts
Comment (secustor) [reply lib/workers/global/autodiscover.ts]: I find this way more readable, but I understand if you want to keep only the necessary changes.

File: lib/workers/global/autodiscover.ts
Comment (secustor) [reply lib/workers/global/autodiscover.ts:91]: I have planned to introduce a `autodiscoverFilterExclude`, which will remove entries after the fact, when this is merged.      
This hadn't been added here to prevent this PR become to bloated, but we can add it together if it is fine for you.

From a UX perpesctive I think it is better to not give the question mark a special meaning which differs from other parts of Renovate.

File: lib/workers/global/config/parse/env.ts
Comment (secustor) [reply lib/workers/global/config/parse/env.ts:168]: This has been part or the original PR, but I going to create a new PR for this.

File: docs/usage/self-hosted-configuration.md
Comment (secustor) [reply docs/usage/self-hosted-configuration.md:90]: Thx, sounds better ^^

File: lib/workers/global/autodiscover.ts
Comment (viceice) [reply lib/workers/global/autodiscover.ts:91]: Ok, but we need to document, that the negative regex don't work as expected when multiple filters are used .

**Review by viceice — COMMENTED** [review]
see comment on conversation

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:90]: Do you mean commas within your filters? Needs clarification, maybe an example

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:110]:
```suggestion
You can negate the regex by putting a `!` in front.
If using negations, all repositories expect those who match the regex are added to the result:
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md:110]: > If using negations, all repositories expect those who match the regex are added to the result:

This is hard to understand for me. How can a repository "expect" things? I think this needs more work. 😉

Comment (secustor) [reply docs/usage/self-hosted-configuration.md]: I have meant `except`

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:90]:
```suggestion
If you use an environment variable or the CLI to set the value for `autodiscoverFilter`, then commas `,` within filters are not supported.
```

File: lib/workers/global/autodiscover.ts
Comment (rarkins) [reply lib/workers/global/autodiscover.ts:91]: Now added to documentation. Would be nice if we can include this as part of v33 though

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:120]:
```suggestion
Only use a single negation and don't mix with other filters because all filters are combined with `or`.
If using negations, all repositories except those who match the regex are added to the result:
```
i think we need to clarify behavior here.

**Review by viceice — COMMENTED** [review]
maybe we need to clarify the `or` behavior once more at beginning of the filter docs?

@HonkingGoose @rarkins WDYT

**Comment by rarkins** [comment]
Does it even make sense to have more than one negation?

**Comment by viceice** [comment]
no, but users tend to think they can do it now. so we should document it's not possible.

**Comment by ab-ty** [comment]
I'm currently trying to use the list and the renovate-config-validator is saying that the array on my json file should be a string.
```
ERROR: config.json contains errors
       "errors": [
         {
           "topic": "Configuration Error",
           "message": "Configuration option `autodiscoverFilter` should be a string"
         }
       ]

```

Is this a problem with the validator, my config or just not available yet?

**Comment by viceice** [comment]
this isn't yet released, so you can't use it.

**Review by rarkins — APPROVED** [review]

**Comment by ab-ty** [comment]
Is this gonna get merged anytime soon?
I'm currently trying to set up a pretty big deployment of renovate and would like to use this feature asap

**Review by viceice — APPROVED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md]:
```suggestion
If you use an environment variable or the CLI to set the value for `autodiscoverFilter`, then you must avoid commas `,` within filters.
```

**Comment by secustor** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}                                                                                                                                               Gesendet über BlackBerry Hub+ Posteingang für Android                                                                                                                                                                           Von: ***@***.***: 7. Oktober 2022 10:26An: ***@***.***: ***@***.***: [GitHub] @n26-terraform-user has invited you to join the @n26 organization  

          
            
               

              
                    
                    
                        @n26-terraform-user has invited you to join the @n26 organization

                    
                  
               


            
                  
                        
                          
  
   

@n26-terraform-user has invited you to join the n26 organization
 




  Hi Sebastian Poxhofer!


  @n26-terraform-user has invited you to join the @n26 organization on GitHub. Head over to https://github.com/n26 to check out @n26’s profile.

  
    This invitation will expire in 7 days.
  


  
      
              Join @n26
          
    



 



  Note: If you get a 404 page, make sure you’re signed in as secustor. You can also accept the invitation by visiting the organization page directly at https://github.com/n26.
    If @n26-terraform-user is sending you too many emails, you can block them or report them for abuse.



  Button not working? Paste the following link into your browser:
  https://github.com/orgs/n26/invitation?via_email=1




                      
                


            
               

                  
                      Opt out of future invitations from this organization ・
                      Manage your GitHub email preferences ・
                      Terms ・
                      Privacy ・
                      Sign in to GitHub 
                  
               

                You’re receiving this email because @n26-terraform-user invited you to an organization on GitHub.


            
   

  GitHub, Inc. ・88 Colin P Kelly Jr Street ・San Francisco, CA 94107

**Comment by viceice** [comment]
??

**Comment by secustor** [comment]
My mail client autocompleted to an Github reply mail address. :/

**Pull request merged**
